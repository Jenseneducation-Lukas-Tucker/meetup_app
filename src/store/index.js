import Vue from 'vue'
import Vuex from 'vuex'
import * as firebase from 'firebase'

Vue.use(Vuex)

export const store = new Vuex.Store({
  state:{
    loadedMeetups: [
      { imageUrl: 'https://cbsnews3.cbsistatic.com/hub/i/r/2020/08/05/e75c7d31-b8e5-4957-b8e6-672664837164/thumbnail/1200x630g8/247b3168454fcbd27326be7ea56d6e98/gettyimages-1208986924.jpg', 
      id: 'asdfdgg', 
      title: 'Meetup in New York',
      date: new Date(),
      location: 'New York',
      description:'NEWWWWWW YOOOORK BABIEE'
    },
      { imageUrl: 'https://cdn-image.travelandleisure.com/sites/default/files/styles/1600x1000/public/1444253482/DG2015-new-york.jpg?itok=neFmsUY1',
      id: 'asdfdpg', 
      title: 'Meetup in Paris',
      date: new Date(),
      location: 'Paris',
      description:'It is Paris!'
    }
    ],
    user: null
  },
  mutations:{
    createMeetup (state, payload) {
      state.loadedMeetups.push(payload)
    },
    setUser (state, payload) {
      state.user = payload
    }
  },
  actions:{
    createMeetup({commit}, payload){
      const meetup = {
        title: payload.title,
        location: payload.location,
        imageUrl: payload.imageUrl,
        description: payload.description,
        date: payload.date,
        id: 'jasbfjasf'
      }
      //reach out to firebase and store it
      commit('createMeetup', meetup)
    },
    signUserUp ({commit}, payload) {
      firebase.auth().createUserWithEmailAndPassword(payload.email, payload.password)
      .then(
        user => {
          const newUser = {
            id: user.uid,
            registeredMeetups: []
          }
          commit('setUser', newUser)
        }
      )
        .catch(
          error => {
            console.log(error)
          }
        )
    },
    signUserIn ({commit}, payload) {
      firebase.auth().signInWithEmailAndPassword(payload.email, payload.password)
      .then(
        user => {
          const newUser = {
            id: user.uid,
            registeredMeetups: []
          }
          commit('setUser', newUser)
        }
      )
      .catch(
        error => {
          console.log(error)
        }
      )
    }
  },
  getters:{
    loadedMeetups (state) {
      return state.loadedMeetups.sort((meetupA, meetupB) => {
        return meetupA.date > meetupB.date
      })
    },
    featuredMeetups (state, getters) {
    return getters.loadedMeetups.slice(0,5)
    },
    loadedMeetup (state) {
      return (meetupId) => {
        return state.loadedMeetups.find((meetup) => {
          return meetup.id === meetupId
        })
      }
    },
    user (state) {
      return state.user
    }
  }
})