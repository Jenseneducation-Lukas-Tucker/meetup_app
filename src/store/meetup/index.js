import Vue from 'vue'
import Vuex from 'vuex'
import * as firebase from 'firebase'

Vue.use(Vuex)

export default {
  state:{
    loadedMeetups: [
      { imageUrl: 'https://cbsnews3.cbsistatic.com/hub/i/r/2020/08/05/e75c7d31-b8e5-4957-b8e6-672664837164/thumbnail/1200x630g8/247b3168454fcbd27326be7ea56d6e98/gettyimages-1208986924.jpg', 
      id: 'asdfdgg', 
      title: 'Meetup in New York',
      date: new Date(),
      location: 'New York',
      description:'NEWWWWWW YOOOORK BABIEE'
    },
      { imageUrl: 'https://cdn-image.travelandleisure.com/sites/default/files/styles/1600x1000/public/1444253482/DG2015-new-york.jpg?itok=neFmsUY1',
      id: 'asdfdpg', 
      title: 'Meetup in Paris',
      date: new Date(),
      location: 'Paris',
      description:'It is Paris!'
    }
    ],
  },
  mutations:{
    setLoadedMeetups (state, payload) {
      state.loadedMeetups = payload
    },
    createMeetup (state, payload) {
      state.loadedMeetups.push(payload)
    },
    updateMeetup(state, payload) {
      const meetup = state.loadedMeetups.find(meetup => {
        return meetup.id === payload.id
      })
      if(payload.title){
        meetup.title = payload.title
      }
      if(payload.description){
        meetup.description = payload.description
      }
      if(payload.date){
        meetup.date = payload.date
      }
    }
  },
  actions:{
    loadMeetups ({commit}){
      commit('setLoading',true)
      firebase.database().ref('meetups').once('value')
      .then((data) => {
        const meetups = []
        const obj = data.val()
        for (let key in obj) {
          meetups.push({
            id: key,
            title: obj[key].title,
            description: obj[key].description,
            imageUrl: obj[key].imageUrl,
            date: obj[key].date,
            location: obj[key].location,
            creatorId: obj[key].creatorId
          })
        }
        commit('setLoadedMeetups', meetups)
        commit('setLoading',false)
      })
      .catch((error)=>{
        console.log(error)
        commit('setLoading',false)
      })
    },
    createMeetup ({commit, getters}, payload) {
      commit('setLoading',true)
      const meetup = {
        title: payload.title,
        location: payload.location,
        description: payload.description,
        date: payload.date.toISOString(),
        creatorId: getters.user.id,
      }
      let imageUrl
      let key
      firebase.database().ref('meetups').push(meetup)
        .then((data) => {
          key = data.key
          return key
        })
        .then(key => {
          const filename = payload.image.name
          const ext = filename.slice(filename.lastIndexOf('.'))
          return firebase.storage().ref('meetups/' + key + ext).put(payload.image)
        })
        .then(fileData => {
          return fileData.ref.getDownloadURL()
          .then(url => {
              imageUrl = url
              return firebase.database().ref('meetups').child(key).update({ imageUrl: url })
          })
        })
        .then(() => {
          commit('createMeetup', {
            ...meetup,
            imageUrl: imageUrl,
            id: key
          })
          commit('setLoading',false)
        })
        .catch((error) => {
          commit('setLoading',false)
          console.log(error)
        })
      // Reach out to firebase and store it
    },
    updateMeetupData({commit}, payload){
      commit('setLoading', true)
      const updateObj = {}
      if (payload.title){
        updateObj.title = payload.title
      }
      if (payload.description){
        updateObj.description = payload.description
      }
      if (payload.date){
        updateObj.date = payload.date
      }
      firebase.database().ref('meetups').child(payload.id).update(updateObj)
      .then(() => {
        commit('setLoading', false)
        commit('updateMeetup', payload)
      })
      .catch(error => {
        console.log(error)
        commit('setLoading', false)
      })
    },
  },
  getters:{
    loadedMeetups (state) {
      return state.loadedMeetups.sort((meetupA, meetupB) => {
        return meetupA.date > meetupB.date
      })
    },
    featuredMeetups (state, getters) {
    return getters.loadedMeetups.slice(0,5)
    },
    loadedMeetup (state) {
      return (meetupId) => {
        return state.loadedMeetups.find((meetup) => {
          return meetup.id === meetupId
        })
      }
    }
  }
}