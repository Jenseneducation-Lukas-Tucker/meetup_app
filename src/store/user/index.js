import Vue from 'vue'
import Vuex from 'vuex'
import * as firebase from 'firebase'

Vue.use(Vuex)

export default {
  state:{
    reviews: [{
      rating: 5,
      review: 'blablabla',
      CreatorId:'sadfgh',
      meetupId:'asdfdgg',
      id:''
    }],
    user: null,
  },
  mutations:{
    setLoadedReviews (state, payload) {
      state.reviews = payload
    },
    registerForMeetup (state, payload){
      const id = payload.id
      if(state.user.registeredMeetups.findIndex(meetup=>meetup.id === id) >= 0){
        return
      }
      state.user.registeredMeetups.push(id)
      state.user.fbKeys[id] = payload.fbKey
    },
    unregisterFromMeetup (state, payload) {
      const registeredMeetups = state.user.registeredMeetups
      registeredMeetups.splice(registeredMeetups.findIndex(meetup=>meetup.id === payload), 1)
      Reflect.deleteProperty(state.user.fbKeys, payload)
    },
    postReview (state, payload){
      state.reviews.push(payload)
    },
    deleteReview(state, payload) {
      const userReviews = state.user.reviews
      userReviews.splice(userReviews.findIndex(meetup=>meetup.id === payload), 1)
      Reflect.deleteProperty(state.user.fbKeys, payload)
    },
    setUser (state, payload) {
      state.user = payload
    }
  },
  actions:{
    loadReviews ({commit, getters}){
      commit('setLoading',true)
      const user = getters.user
      firebase.database().ref('/users/' + user.id + '/reviews/').once('value')
      .then((data) => {
        const reviews = []
        const object = data.val()
        console.log(object)

        for (let key in object) {
            reviews.push({
            id: key,
            rating: object[key].rating,
            review: object[key].review,
            creatorId: object[key].creatorId,
            meetupId: object[key].meetupId
          })
        }
        commit('setLoadedReviews', reviews)
        commit('setLoading',false)
      })
      .catch((error)=>{
        console.log(error)
        commit('setLoading',false)
      })
    },
    registerForMeetup({commit, getters}, payload) {
      commit('setLoading', true)
      const user = getters.user
      firebase.database().ref('/users/' + user.id).child('/registrations/')
      .push(payload)
      .then(data => {
        commit('setLoading', false)
        commit('registerForMeetup', {id:payload, fbKey:data.key})
      })
      .catch(error => {
        console.log(error)
        commit('setLoading', false)
      })
    },
    unregisterFromMeetup({commit, getters}, payload) {
      commit('setLoading', true)
      const user = getters.user
      if (!user.fbKeys){
        return
      }
      const fbKey = user.fbKeys[payload]
      firebase.database().ref('/users/' + user.id + '/registrations/').child(fbKey)
        .remove()
      .then(() => {
        commit('setLoading', false)
        commit('unregisterFromMeetup', payload)
      })
      .catch(error => {
        console.log(error)
        commit('setLoading',false)
      })
    },
    postReview({commit, getters}, payload) {
      commit('setLoading', true)
      const user = firebase.auth().currentUser
      const review = {
        rating: payload.rating,
        review: payload.review,
        meetupId: payload.meetupId,
        creatorId: user.email
      }
      firebase.database().ref('/users/' + getters.user.id).child('/reviews/').push(review)
      .then((data) =>{ 
        const key = data.key
        commit('setLoading', false)
        console.log(data)
        commit('postReview', {
          ...review,
          id: key})
      })
      .catch(error =>{
        commit('setLoading', false)
        console.log(error)
      })
      },

    deleteReview({commit, getters},payload){
      commit('setLoading', true)
      const user = getters.user
      if (!user.fbKeys){
        return
      }
      const fbKey = user.fbKeys[payload]
      firebase.database().ref('/users/' + user.id + '/reviews/').child(fbKey)
        .remove()
      .then(() => {
        commit('setLoading', false)
        commit('unregisterFromMeetup', payload)
      })
      .catch(error => {
        console.log(error)
        commit('setLoading',false)
      })
    },
    signUserUp ({commit}, payload) {
      commit('setLoading',true)
      commit('clearError')
      firebase.auth().createUserWithEmailAndPassword(payload.email, payload.password)
      .then(
        user => {
          commit('setLoading', false)
          const newUser = {
            id: user.uid,
            registeredMeetups: [],
            reviews: [],
            userReviews: [],
            fbKeys: {}
          }
          commit('setUser', newUser)
        }
      )
        .catch(
          error => {
            commit('setLoading', false)
            commit('setError', error)
            console.log(error)
          }
        )
    },
    signUserIn ({commit}, payload) {
      commit('setLoading',true)
      commit('clearError')
      firebase.auth().signInWithEmailAndPassword(payload.email, payload.password)
      .then(
        user => {
          commit('setLoading', false)
          const newUser = {
            id: user.uid,
            registeredMeetups: [],
            reviews: [],
            userReviews: [],
            fbKeys: {}
          }
          commit('setUser', newUser)
        }
      )
      .catch(
        error => {
          commit('setLoading', false)
          commit('setError', error)
          console.log(error)
        }
      )
    },
    autoSignIn({commit}, payload) {
      commit('setUser', {
        id:payload.uid,
        registeredMeetups: [],
        fbKeys: {}
      })
    },
    fetchUserData ({commit, getters}) {
      commit('setLoading', true)
      firebase.database().ref('/users/' + getters.user.id + '/registrations/').once('value')
      .then(data => {
        const dataPairs = data.val()
        let registeredMeetups = []
        let swappedPairs = {}
        for (let key in dataPairs) {
          registeredMeetups.push(dataPairs[key])
          swappedPairs[dataPairs[key]] = key
        }
        const updatedUser = {
          id: getters.user.id,
          registeredMeetups: registeredMeetups,
          fbKeys: swappedPairs
        }
        commit('setLoading', false)
        commit('setUser', updatedUser)
      })
      .catch(error => {
        console.log(error)
        commit('setLoading',false)
      })
    },
      logout({commit}){
        firebase.auth().signOut()
        commit('setUser', null)
      }
  },
  getters:{
    loadedReviews (state) {
      return state.reviews.sort((reviewA, reviewB) => {
        return reviewA.rating > reviewB.rating
      })
      },
    user (state) {
      return state.user
    }
  }
}