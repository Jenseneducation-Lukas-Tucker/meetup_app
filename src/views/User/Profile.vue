<template>
<div class="text-center">
  <h1>Coming Soon!</h1>
  <h2 class="primary--text">Organize your meetup in the meantime!</h2>
  </div>
</template>

<script>
export default {
}
</script>

<style>
h2 :hover{
cursor: pointer;
}
</style>