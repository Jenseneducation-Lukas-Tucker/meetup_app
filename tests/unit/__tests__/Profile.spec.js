import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
//Libraries
import Vuex from 'vuex';
import Vue from 'vue'
import Vuetify from 'vuetify'
//store
import store from '@/store/user/index.js';
//views
import Profile from '@/views/User/Profile.vue';
import VueRouter from 'vue-router';

const localVue = createLocalVue();
localVue.use(Vuex, VueRouter);

describe('CreateMeetup.vue', () => {
  let localVue;
  let vuetify;
  let wrapper;
  beforeEach(() => {
    localVue = createLocalVue(); // because of vuetify, we should use a localVue instance
    vuetify = new Vuetify();
    wrapper = mount(Profile, {
      localVue,
      vuetify,
    });
  });
test('It renders correctly and matches snapshot', async () => {
  const wrapper = mount(Profile, {
    store,
    localVue,
    vuetify,
  });
  expect(wrapper.html()).toMatchSnapshot();
});
});