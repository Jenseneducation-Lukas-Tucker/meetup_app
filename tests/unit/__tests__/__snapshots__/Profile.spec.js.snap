// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMeetup.vue It renders correctly and matches snapshot 1`] = `
<div class="text-center">
  <h1>Coming Soon!</h1>
  <h2 class="primary--text">Organize your meetup in the meantime!</h2>
</div>
`;
